core/log/performance_hint.cpp

```diff
@@ -0,0 +1,190 @@
+    }
+    GKO_ASSERT(map.size() == entries.size());
+    map.clear();
+    std::sort(entries.begin(), entries.end(),
+              [](auto a, auto b) { return a.second > b.second; });
+    GKO_ASSERT(target_size < max_size);
+    GKO_ASSERT(target_size <= entries.size());
+    entries.erase(entries.begin() + target_size, entries.end());
+    map.insert(entries.begin(), entries.end());
+}
+
+
+void print_allocation_message(std::ostream& stream, size_type size, int count)
+{
+    stream << "Observed " << count << " allocate-free pairs of size " << size
+           << " that may point to unnecessary allocations.\n";
+}
+
+
+void print_copy_from_message(std::ostream& stream, gko::uintptr ptr, int count)
+{
+    stream << "Observed " << count << " cross-executor copies from "
+           << location_name(ptr)
+           << " that may point to unnecessary data transfers.\n";
+}
+
+
+void print_copy_to_message(std::ostream& stream, gko::uintptr ptr, int count)
+{
+    stream << "Observed " << count << " cross-executor copies to "
+           << location_name(ptr)
+           << " that may point to unnecessary data transfers.\n";
+}
+
+
+}  // namespace
+
+
+void PerformanceHint::on_allocation_completed(const Executor* exec,
+                                              const size_type& num_bytes,
+                                              const uintptr& location) const
+{
+    if (num_bytes > allocation_size_limit_) {
+        allocation_sizes_[location] = num_bytes;
+        // erase smallest allocations first
+        compact_storage(allocation_sizes_, histogram_max_size,
+                        histogram_max_size * 3 / 4);
+    }
+}
+
+
+void PerformanceHint::on_free_completed(const Executor* exec,
+                                        const uintptr& location) const
+{
+    const auto it = allocation_sizes_.find(location);
+    if (it != allocation_sizes_.end()) {
+        const auto size = it->second;
+        allocation_sizes_.erase(it);
+        const auto count = ++allocation_histogram_[size];
+        if (should_log_count(count)) {
+            print_allocation_message(log(), size, count);
+        }
+        // erase rarest allocation sizes first
+        compact_storage(allocation_histogram_, histogram_max_size,
+                        histogram_max_size * 3 / 4);
+    }
+}
+
+
+void PerformanceHint::on_copy_completed(const Executor* from,
+                                        const Executor* to,
+                                        const uintptr& location_from,
+                                        const uintptr& location_to,
+                                        const size_type& num_bytes) const
+{
+    if (num_bytes > copy_size_limit_ && from != to) {
+        const auto count1 = ++copy_src_histogram_[location_from];
+        const auto count2 = ++copy_dst_histogram_[location_to];
+        if (should_log_count(count1)) {
+            print_copy_from_message(log(), location_from, count1);
```

Comment:
I think it might be useful to also print the executor type ? Might be a bit ugly to do dynamic casts to all the concrete executor types, but I think that might be useful information.

This might also be a time to add a get executor string method to the base Executor class and have the concrete classes store and return the names.

Reply:
name_demangling should work there :)

Reply:
actually it doesn't work too well, since I don't keep the information which allocator a pointer is associated with, and this doesn't work with `print_status`. Storing the executor would work for copies, but not for allocations.

Reply:
But you do have the `from` and `to` executors available in the function, that you can propagate to the print functions ?

Reply:
they are no longer available in the print_status function though, which is the only issue here